### PR TITLE
feat: allow more flexible http requests in updater closes #7422

### LINKED ISSUE #7422: [feat] More flexibility in HTTP requests used by Updater
### Describe the problem

We need to be able to construct different headers for the HTTP request used to check for available updates, and the HTTP  request used to download the update. 

For example, if you set AWS SIGv4 headers on the request to check for updates, the same headers are not appropriate to attempt to download the update from AWS S3 (in fact the S3 request is impossible) - but I can find no way to modify the headers between checking and downloading.  

Sending signed requests, using user provided credentials for example,  would also be significantly easier if the updater endpoints could be modified after application start (different from the solution suggested https://github.com/tauri-apps/tauri/issues/2595#issuecomment-1160772317) 

### Describe the solution you'd like

Minimally, setters on `UpdateResponse` that allowed the headers to be changed. 
Ideally, also setters on `UpdateBuilder` that allowed the endpoints to be changed dynamically. 

### Alternatives considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by gmchale79**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x ] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [ ] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### Other information
These small changes allow us to use AWS SigV4 authenticated requests for checking for updates, and pre-signed S3 URLs for updater downloads - a call flow that was previously not possible. 
I'm aware that there is other work in this area in 2.x, but we need to use 1.x for the foreseeable future.

### CODE REVIEW

File: core/tauri/src/updater/core.rs [hunk 1]
```diff
@@ -505,6 +505,33 @@ impl<R: Runtime> Clone for Update<R> {
...
+  where
+    HeaderName: TryFrom<K>,
+    <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
+  {
+    let key: std::result::Result<HeaderName, http::Error> = key.try_into().map_err(Into::into);
+    self.headers.remove(key?);
+    Ok(self)
+  }
+}
+
```
Comment (amrbashir) [inline core/tauri/src/updater/core.rs]: why open a new `impl` block? just reuse the one below on line 535

File: core/tauri/src/updater/core.rs [hunk 2]
```diff
@@ -505,6 +505,33 @@ impl<R: Runtime> Clone for Update<R> {
   }
 }
 
+impl<R: Runtime> Update<R> {
+  pub fn header<K, V>(mut self, key: K, value: V) -> Result<Self>
```
Comment (amrbashir) [inline core/tauri/src/updater/core.rs:505]:
```suggestion
  pub(crate) fn header<K, V>(mut self, key: K, value: V) -> Result<Self>
```

File: core/tauri/src/updater/mod.rs [hunk 3]
```diff
@@ -275,6 +275,12 @@ impl<R: Runtime> UpdateBuilder<R> {
     Ok(self)
   }
 
+  /// Add endpoints.
```
Comment (amrbashir) [inline core/tauri/src/updater/mod.rs]:
```suggestion
  /// Set the endpoints.
```

File: core/tauri/src/updater/mod.rs [hunk 4]
```diff
@@ -395,6 +402,28 @@ impl<R: Runtime> UpdateResponse<R> {
...
+
+  /// Removes a header from the update request.
+  pub fn remove_header<K>(mut self, key: K) -> Result<Self>
+  where
+    HeaderName: TryFrom<K>,
+    <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
+  {
+    self.update = self.update.remove_header(key)?;
+    Ok(self)
+  }
```
Comment (amrbashir) [inline core/tauri/src/updater/mod.rs]: would you explain why would you need this?

**Review by amrbashir — CHANGES_REQUESTED** [review]
Thank you for the contribution, this looks okay but just needs a few modifications.

Also please add a change file in `.changes` directory

File: core/tauri/src/updater/mod.rs [hunk 5]
```diff
@@ -275,6 +275,12 @@ impl<R: Runtime> UpdateBuilder<R> {
     Ok(self)
   }
 
+  /// Add endpoints.
+  pub fn endpoints(mut self, urls: &[String]) -> Result<Self> {
+    self.inner = self.inner.urls(urls);
+    Ok(self)
+  }
```
Comment (amrbashir) [inline core/tauri/src/updater/mod.rs]:
```suggestion
  pub fn endpoints(mut self, urls: &[String]) -> Self {
    self.inner = self.inner.urls(urls);
    self
  }
```

File: core/tauri/src/updater/mod.rs (hunk 4 above)
Comment (lucasfernog) [reply core/tauri/src/updater/mod.rs]: probably because he needs a header for the JSON fetch request, but not for the install request

**Review by lucasfernog — APPROVED** [review]

**Comment by gmchale79** [comment]
Yes, exactly. If we set signed headers for the fetch request, they are not valid for the subsequent install request to S3 and S3 rejects them. Apologies for the delayed response.

File: core/tauri/src/updater/mod.rs (hunk 4 above)
Comment (gmchale79) [reply core/tauri/src/updater/mod.rs]: Yes, exactly. If we set signed headers for the fetch request, they are not valid for the subsequent install request to S3 and S3 rejects them. Apologies for the delayed response.

**Review by amrbashir — APPROVED** [review]

**Pull request merged**